### PR TITLE
Implement open telemetry integration - client/server and explict token latency

### PR DESCRIPTION
**Pull request by ppalucki**
## Description

Integrate OpenTelemetry (**metrics only**) and expose it using Promteheus exporter for:

1. Server (incomming connection) (request/reponse size, duration) with **http_route** label for each handler:
   - **http_server_duration_milliseconds** (histogram: bucket/total/sum)
   - **http_server_request_size_bytes_total** (counter)
   - **http_server_response_size_bytes_total** (counter)
2. Client (all outgoing connection to services) with **net_peer_name** label to distinguish target:
   - **http_client_duration_milliseconds** (histogram: bucket/total/sum)
   - **http_client_request_size_bytes_total** (counter)
   - **http_client_response_size_bytes_total** (counter)
3. Explict Float64 historgrams for first and all next tokens lantecy:
   - **llm_first_token_latency_milliseconds** (histogram: bucket/total/sum)
   - **llm_next_token_latency_milliseconds** (histogram: bucket/total/sum)
   - **llm_all_token_latency_milliseconds** (histogram: bucket/total/sum)
   - **llm_pipeline_latency_milliseconds** (histogram: bucket/total/sum)
4. Other default Go language metrics like: **go_gc_duration_seconds**, **go_goroutines** and others.

Additional feature is integrated opentelemetry trace context propagation (**not tested**).

Metrics are exposed on /metric with Prometheus handler. Example output for metrics:

#### 1) server metrics example
```
# HELP http_server_duration_milliseconds Measures the duration of inbound HTTP requests.
# TYPE http_server_duration_milliseconds histogram
http_server_duration_milliseconds_bucket{http_method="POST",http_route="/",http_scheme="http",http_status_code="200",net_host_name="127.0.0.1",net_host_port="8080",net_protocol_name="http",net_protocol_version="1.1",otel_scope_name="go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",otel_scope_version="0.49.0",le="0"} 0
http_server_duration_milliseconds_bucket{http_method="POST",http_route="/",http_scheme="http",http_status_code="200",net_host_name="127.0.0.1",net_host_port="8080",net_protocol_name="http",net_protocol_version="1.1",otel_scope_name="go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",otel_scope_version="0.49.0",le="5"} 0
http_server_duration_milliseconds_bucket{http_method="POST",http_route="/",http_scheme="http",http_status_code="200",net_host_name="127.0.0.1",net_host_port="8080",net_protocol_name="http",net_protocol_version="1.1",otel_scope_name="go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",otel_scope_version="0.49.0",le="+Inf"} 2
http_server_duration_milliseconds_sum{http_method="POST",http_route="/",http_scheme="http",http_status_code="200",net_host_name="127.0.0.1",net_host_port="8080",net_protocol_name="http",net_protocol_version="1.1",otel_scope_name="go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",otel_scope_version="0.49.0"} 25597.877786
http_server_duration_milliseconds_count{http_method="POST",http_route="/",http_scheme="http",http_status_code="200",net_host_name="127.0.0.1",net_host_port="8080",net_protocol_name="http",net_protocol_version="1.1",otel_scope_name="go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",otel_scope_version="0.49.0"} 2
# HELP http_server_request_size_bytes_total Measures the size of HTTP request messages.
# TYPE http_server_request_size_bytes_total counter
http_server_request_size_bytes_total{http_method="POST",http_route="/",http_scheme="http",http_status_code="200",net_host_name="127.0.0.1",net_host_port="8080",net_protocol_name="http",net_protocol_version="1.1",otel_scope_name="go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",otel_scope_version="0.49.0"} 202
# HELP http_server_response_size_bytes_total Measures the size of HTTP response messages.
# TYPE http_server_response_size_bytes_total counter
http_server_response_size_bytes_total{http_method="POST",http_route="/",http_scheme="http",http_status_code="200",net_host_name="127.0.0.1",net_host_port="8080",net_protocol_name="http",net_protocol_version="1.1",otel_scope_name="go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",otel_scope_version="0.49.0"} 134
```

#### 2) client metrics example
```
# HELP http_client_duration_milliseconds Measures the duration of outbound HTTP requests.
# TYPE http_client_duration_milliseconds histogram
http_client_duration_milliseconds_bucket{http_method="POST",http_status_code="200",net_peer_name="embedding-svc.chatqa.svc.cluster.local",net_peer_port="6000",otel_scope_name="go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",otel_scope_version="0.49.0",le="5000"} 2
...
http_client_duration_milliseconds_bucket{http_method="POST",http_status_code="200",net_peer_name="llm-svc.chatqa.svc.cluster.local",net_peer_port="9000",otel_scope_name="go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",otel_scope_version="0.49.0",le="0"} 0
...
# HELP http_client_request_size_bytes_total Measures the size of HTTP request messages.
# TYPE http_client_request_size_bytes_total counter
http_client_request_size_bytes_total{http_method="POST",http_status_code="200",net_peer_name="embedding-svc.chatqa.svc.cluster.local",net_peer_port="6000",otel_scope_name="go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",otel_scope_version="0.49.0"} 202
http_client_request_size_bytes_total{http_method="POST",http_status_code="200",net_peer_name="llm-svc.chatqa.svc.cluster.local",net_peer_port="9000",otel_scope_name="go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",otel_scope_version="0.49.0"} 3670
http_client_request_size_bytes_total{http_method="POST",http_status_code="200",net_peer_name="reranking-svc.chatqa.svc.cluster.local",net_peer_port="8000",otel_scope_name="go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",otel_scope_version="0.49.0"} 14076
http_client_request_size_bytes_total{http_method="POST",http_status_code="200",net_peer_name="retriever-svc.chatqa.svc.cluster.local",net_peer_port="7000",otel_scope_name="go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",otel_scope_version="0.49.0"} 19498
# HELP http_client_response_size_bytes_total Measures the size of HTTP response messages.
# TYPE http_client_response_size_bytes_total counter
http_client_response_size_bytes_total{http_method="POST",http_status_code="200",net_peer_name="embedding-svc.chatqa.svc.cluster.local",net_peer_port="6000",otel_scope_name="go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",otel_scope_version="0.49.0"} 19422
http_client_response_size_bytes_total{http_method="POST",http_status_code="200",net_peer_name="llm-svc.chatqa.svc.cluster.local",net_peer_port="9000",otel_scope_name="go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",otel_scope_version="0.49.0"} 480
http_client_response_size_bytes_total{http_method="POST",http_status_code="200",net_peer_name="reranking-svc.chatqa.svc.cluster.local",net_peer_port="8000",otel_scope_name="go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",otel_scope_version="0.49.0"} 3640
http_client_response_size_bytes_total{http_method="POST",http_status_code="200",net_peer_name="retriever-svc.chatqa.svc.cluster.local",net_peer_port="7000",otel_scope_name="go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp",otel_scope_version="0.49.0"} 13972
```

#### 3) custom metrics example
```
# HELP llm_first_token_latency_milliseconds Measures the first token generate duration of.
# TYPE llm_first_token_latency_milliseconds histogram
llm_first_token_latency_milliseconds_bucket{otel_scope_name="entrag-telemetry",otel_scope_version="",le="1"} 0
llm_first_token_latency_milliseconds_bucket{otel_scope_name="entrag-telemetry",otel_scope_version="",le="64"} 0
...
llm_first_token_latency_milliseconds_bucket{otel_scope_name="entrag-telemetry",otel_scope_version="",le="+Inf"} 2
llm_first_token_latency_milliseconds_sum{otel_scope_name="entrag-telemetry",otel_scope_version=""} 9550.203300000001
llm_first_token_latency_milliseconds_count{otel_scope_name="entrag-telemetry",otel_scope_version=""} 2
# HELP llm_next_token_latency_milliseconds Measures the next token generate duration of.
# TYPE llm_next_token_latency_milliseconds histogram
llm_next_token_latency_milliseconds_bucket{otel_scope_name="entrag-telemetry",otel_scope_version="",le="1"} 4
llm_next_token_latency_milliseconds_bucket{otel_scope_name="entrag-telemetry",otel_scope_version="",le="64"} 4
...
llm_next_token_latency_milliseconds_bucket{otel_scope_name="entrag-telemetry",otel_scope_version="",le="+Inf"} 36
llm_next_token_latency_milliseconds_sum{otel_scope_name="entrag-telemetry",otel_scope_version=""} 14365.578998999998
llm_next_token_latency_milliseconds_count{otel_scope_name="entrag-telemetry",otel_scope_version=""} 36
```

#### 4) Prometheus client default Go language metrics:
```
skl212 ➜  ~ curl -sL -N http://127.0.0.1:8080/metrics G go_ | grep -v HELP | grep -v TYPE
go_gc_duration_seconds{quantile="0"} 9.7515e-05
go_gc_duration_seconds{quantile="0.25"} 0.000102628
go_gc_duration_seconds{quantile="0.5"} 0.000134259
go_gc_duration_seconds{quantile="0.75"} 0.000141445
go_gc_duration_seconds{quantile="1"} 0.000191476
go_gc_duration_seconds_sum 0.00130944
go_gc_duration_seconds_count 10
go_goroutines 8
go_info{version="go1.22.4"} 1
go_memstats_alloc_bytes 2.795128e+06
....
go_memstats_sys_bytes 2.276456e+07
go_threads 23
```

## Issues

n/a

## Type of change

- [x] New feature (non-breaking change which adds new functionality)

## Dependencies

- open telemetry go sdk (+prometheus exporter)
   - `go.opentelemetry.io/otel`
   - `go.opentelemetry.io/otel/exporters/prometheus`
   - `go.opentelemetry.io/otel/metric`
   - `go.opentelemetry.io/otel/sdk/metric`
   - `go.opentelemetry.io/otel/metric`
- prometheus client:
   - `github.com/prometheus/client_golang/prometheus/promhttp`
- open telemetry go contrib http client/server intrumentation 
   - `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp`

## Tests

Manually using curl (examples show above) after running only with ChatQnA sample graph.

## TODO:

- [ ] better or new label and attributes (e.g. otel_scope_name...)
- [ ] better buckets boundries for token latencies
- [ ] reconfigure default bucket boundries for server/client latnecy metrics
- [ ] test OTEL trace propagation
- [ ] write unit tests
- [ ] document
- [x] remove some comments with examples

### CODE REVIEW

File: microservices-connector/cmd/router/main.go [hunk 1]
```diff
@@ -517,7 +611,14 @@ func routeStep(nodeName string,
...
 func mcGraphHandler(w http.ResponseWriter, req *http.Request) {
-	ctx, cancel := context.WithTimeout(req.Context(), time.Minute)
+
+	// TODO: traces
+	// span := trace.SpanFromContext(ctx)
+	// bag := baggage.FromContext(ctx)
+	// span.AddEvent("handling this...", trace.WithAttributes(uk.String(bag.Member("username").Value())))
+
+	_, _ = io.WriteString(w, "Hello, world!\n")
+	ctx, cancel := context.WithTimeout(req.Context(), 3600*time.Minute)
```
Comment (zhlsunshine) [inline microservices-connector/cmd/router/main.go:621]: Hi @ppalucki,  may 1 hour time out be too long?

**Comment by zhlsunshine** [comment]
Hi @ppalucki, thanks for your effort on this PR! 
In `mcGraphHandler` : 
```
responseBody, statusCode, err := routeStep(defaultNodeName, *mcGraph, inputBytes, inputBytes, req.Header)
```
I think there should be a measurement after `routeStep` returns the `responseBody` because this produce contains the duration for going through all microservice in the pipeline. And this information can calculate and evaluate the handling time for GMC router. It really is helpful for GMC router. 

I think you also can support the measurement metrics in `mcDataHandler` which is focusing on data embedding and storage in `VectorDB`.

File: microservices-connector/cmd/router/main.go [hunk 2]
```diff
@@ -726,8 +840,26 @@ func handleMultipartError(writer *multipart.Writer, err error) {
...
+
+	handleFunc("/", mcGraphHandler, "mcGraphHandler")
+	handleFunc("/dataprep", mcDataHandler, "mcDataHandler")
+
+	// TODO: what is option as a string?!?!
+	//mux.Handle("/", otelhttp.NewHandler(http.HandlerFunc(mcGraphHandler), "mcGraphHandler"))
+	//mux.Handle("/dataprep", otelhttp.NewHandler(http.HandlerFunc(mcDataHandler), "mcDataHandler"))
+
+	// Export metric using Prometheus client built-in handler.
+	mux.Handle("/metrics", promhttp.Handler())
```
Comment (zhlsunshine) [inline microservices-connector/cmd/router/main.go:861]: Hi @ppalucki, is it possible to registry this router `metrics`  via function `handleFunc` as well?

File: microservices-connector/cmd/router/main.go [hunk 3]
```diff
@@ -517,7 +611,14 @@ func routeStep(nodeName string,
...
 
 func mcGraphHandler(w http.ResponseWriter, req *http.Request) {
-	ctx, cancel := context.WithTimeout(req.Context(), time.Minute)
+
+	// TODO: traces
+	// span := trace.SpanFromContext(ctx)
+	// bag := baggage.FromContext(ctx)
+	// span.AddEvent("handling this...", trace.WithAttributes(uk.String(bag.Member("username").Value())))
+
+	_, _ = io.WriteString(w, "Hello, world!\n")
```
[the remark this answers is not in the document]
Comment (ppalucki) [reply microservices-connector/cmd/router/main.go:620]: @zhlsunshine I accidently copied it from some working branch of streaming, later rebased - now it is removed.

File: microservices-connector/cmd/router/main.go (hunk 1 above)
Comment (ppalucki) [reply microservices-connector/cmd/router/main.go:621]: @zhlsunshine fixed (long timeout was used it for my debugging sessions)

File: microservices-connector/cmd/router/main.go (hunk 2 above)
Comment (ppalucki) [reply microservices-connector/cmd/router/main.go:861]: @zhlsunshine fixed
yes, it is possible, but didn't wrappe it on purpose to exclude from metrics 

but you're right - it should be covered and exlucded later in visualization/post processing

**Comment by ppalucki** [comment]
> Hi @ppalucki, thanks for your effort on this PR! In `mcGraphHandler` :
> 
> ```
> responseBody, statusCode, err := routeStep(defaultNodeName, *mcGraph, inputBytes, inputBytes, req.Header)
> ```
> 
> I think there should be a measurement after `routeStep` returns the `responseBody` because this produce contains the duration for going through all microservice in the pipeline. And this information can calculate and evaluate the handling time for GMC router. It really is helpful for GMC router.
> 
> I think you also can support the measurement metrics in `mcDataHandler` which is focusing on data embedding and storage in `VectorDB`.

I quickly added `all_token_latency`, which is essenatialy sum of first and "all"  next tokens, but still not sure that it covers what you wanted e.g. "user facing" latency of all steps in pipeline + network.

I need to test it more carefully.

```
llm_all_token_latency_milliseconds_sum{otel_scope_name="entrag-telemetry",otel_scope_version=""} 217115.895284
llm_first_token_latency_milliseconds_sum{otel_scope_name="entrag-telemetry",otel_scope_version=""} 85542.49316200001
llm_next_token_latency_milliseconds_sum{otel_scope_name="entrag-telemetry",otel_scope_version=""} 127367.80455300014
```

**Comment by zhlsunshine** [comment]
Hi @ppalucki, I think you may need to fix the DCO error by adding the `Signed-off-by` with your github id and email.

File: microservices-connector/cmd/router/main.go [hunk 4]
```diff
@@ -80,6 +92,69 @@ type ReadCloser struct {
...
+
+	pipelineLatencyMeasure, err = meter.Float64Histogram(
+		"llm.pipeline.latency",
+		metric.WithUnit("ms"),
+		metric.WithDescription("Measures the duration to going through pipeline steps until first token is being generated (including read data time from client)."),
+		api.WithExplicitBucketBoundaries(1, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16364),
+	)
+	if err != nil {
+		log.Error(err, "metrics: cannot register pipeline histogram measure")
+	}
```
[the remark this answers is not in the document]
Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:155]: nice catch!

File: microservices-connector/cmd/router/main.go [hunk 5]
```diff
@@ -80,6 +92,69 @@ type ReadCloser struct {
...
+	// implements prometheus.Collector, allowing it to be used as
+	// both a Reader and Collector.
+	exporter, err := prometheus.New()
+	if err != nil {
+		log.Error(err, "metrics: cannot init prometheus collector")
+	}
+	provider := sdkmetric.NewMeterProvider(sdkmetric.WithReader(exporter))
+	otel.SetMeterProvider(provider)
+
+	// ppalucki: Own metrics defintion bellow
```
[the remark this answers is not in the document]
Comment (zhlsunshine) [reply microservices-connector/cmd/router/main.go:114]: nice catch!

**Comment by eero-t** [comment]
See also: https://github.com/opea-project/GenAIComps/pull/845

**Comment by ppalucki** [comment]
This PR is going to be superseded by new work - please check this #644

Also some issues were found:
- first token latency should be be renamed to TTFT (time-to-first-token) or better "time to first bytes"
- next token latency should be called TPOT (time-per-output-token) or ITL (inter-token-latency)
- first token should measure not just "model server" first token generation but also pipeline latency (before request even is redirected to model llm-server) and first token should reflect that time

Here is proposed change semantic to generated metrics:

Currently (wrong):
```
[ embed]->[retrive]->[reranking]->[vllm] 
                                  "first  token"   "token1"  "token2"

<-------- pipeline latency -->    <-------- all token latency ------>
                                  <-first token->  <--ntl--> <--ntl-->

```
and should be (fixed):
```
[ embed]->[retrive]->[reranking]->[vllm] 
                                  "first  token"   "token1"  "token2"

<------------------------- e2e latency ------------------------------>
<-------- pipeline latency -----> <---------- model server latency -->
<-----------------------------------first token->  <--ntl--> <--ntl-->
```  

where ntl stands for "next token latnecy"

**Pull request closed without merging**